### ISSUE TITLE
(WebVR) Trying out various devices

### ISSUE THREAD

**Issue by Joo918** [x3dom/x3dom#943]
Hello.
I was trying to launch this:
http://metagrid2.sv.vt.edu/~chris582/fish_pool_VR/fish_pool_videosphere_01.html
with these devices:
- Gear VR
- Google Cardboard
- HTC Vive with a wireless adapter
(For Gear VR and Google Cardboard, I used a Samsung Galaxy S7)

When I click on the bottom-left VR button,
it stops at the circular loading bar.
Does anyone else have the same issue?

For the HTV Vive, I got an error on the console saying: "display not recognized"

+ I also tried an A-frame based WebVR application and those go into VR without a problem...

**Comment by andreasplesch** [comment]
This web page does not seem to use the webvr enabled version of x3dom ? You will to need to build from the webvr brnach and use that x3dom-full.js then.

**Comment by andreasplesch** [comment]
Sorry, I was wrong. It does. I do not get the VR button on a Galaxy S6.

**Comment by Joo918** [comment]
Which headset device did you use?

**Comment by andreasplesch** [comment]
cardboard, so just the phone. aframe works with it.

---on the phone---

**Comment by Joo918** [comment]
Yeah..
To get the VR button, I used Yandex Browser to install the WebVR emulator extension from the Chrome store.
Even after getting in VR, it seems that the gyro acceleration sensor input is not being used for the navigation, so I had to drag my fingers on the phone to rotate the camera.

**Comment by andreasplesch** [comment]
It may have to do with possible changes in webvr and chrome.

**Comment by Joo918** [comment]
Hm.. Thanks for the pointer!

**Comment by andreasplesch** [comment]
It is also possible that now an origin-trial token is required for chrome. aframe use this in the html header:
```<meta http-equiv="origin-trial" data-feature="WebVR (For Chrome M59+)" data-expires="2017-07-28" content="ArFv1ZeTwzkhjNE00uAE+XtiQB41fwqG/TqlFMLrepd9sforQSvQE/tgfIbUMYNuNre4QR1k4/z8xp2mV3dbhwwAAABeeyJvcmlnaW4iOiJodHRwczovL2FmcmFtZS5pbzo0NDMiLCJmZWF0dXJlIjoiV2ViVlIxLjEiLCJleHBpcnkiOjE1MDEyMTcwMDIsImlzU3ViZG9tYWluIjp0cnVlfQ==">```

**Comment by andreasplesch** [comment]
Perhaps also a webvr-polyfill is needed:
``` <script src='https://cdn.jsdelivr.net/npm/webvr-polyfill@latest/build/webvr-polyfill.js'></script>```

**Comment by andreasplesch** [comment]
chrome://flags/ webvr enable

**Comment by andreasplesch** [comment]
Enabling webvr in chrome flags gives me now the vr goggles icon and I can switch to vr on Android.

Then I get the same spinner and a timeout error message.

It may have to do with videos only being played after an interaction with the page.

**Comment by andreasplesch** [comment]
chromium-review.googlesource.com/c/chromium/src/+/636386

There is an enforced timeout after 5s if there is no frane rendered. Perhaps this is related.

**Comment by Joo918** [comment]
First of all, thanks for your works over the weekend, @andreasplesch .

The timeout could be the issue, but I'm a bit skeptical..
The application works perfectly fine on an Oculus Go, which is also Android based.
It's also not because the Oculus Browser has something different than Chrome.
I tried the app on a Gear VR which has Oculus Browser, and it also doesn't work.

**Comment by andreasplesch** [comment]
Sure. Do you have an example of a working scene on cardboard ?

**Comment by Joo918** [comment]
I'm afraid not..
Only A-frame based examples worked on a cardboard.
I'm currently trying to see what the console says when I try to go into VR using a Gear VR.

**Comment by andreasplesch** [comment]
I was able to find a Oculus Go, and try out these:
http://metagrid2.sv.vt.edu/~chris582/webVR_demos_tablePage.html
They work pretty well in the Go. But they all have the same timeout issue on cardboard.
I think the only problem with the fishtank after it gets displayed is that when entering VR you get transported outside the Sphere and look at it from the outside rather than the inside.

**Comment by Joo918** [comment]
Thanks for testing.
The reason why you spawn outside the sphere is because a new viewarea is created upon entering VR and it doesn't use the previous viewpoint.
I am also trying to fix that.

**Comment by andreasplesch** [comment]
I think there is another issue for that where we started to discuss navigation in VR.

The core issue is that right now the viewmatrix is taken directly from the headset without taking into account viewpoints at all:

https://github.com/x3dom/x3dom/blob/b90d5cfa2be1d67605cf4a36ad98e6a9cddfffb8/src/Viewarea.js#L498

So the only way right now is to move the world relative to the headset which presumably starts at the origin when entering VR.

You could just play with Viewarea.js a bit:

Perhaps try:

```js
x3dom.Viewarea.prototype.getViewMatrices = function()
{
    if(this.vrFrameData) {
        var viewMatrix = this.getViewpointMatrix();
        return [viewMatrix.mult(this.vrLeftViewMatrix),
                viewMatrix.mult(this.vrRightViewMatrix)];
    } else {
        var viewMatrix = this.getViewpointMatrix().mult(this._transMat).mult(this._rotMat);
        return [viewMatrix, viewMatrix]
    }
};
```
and similar for x3dom.Viewarea.prototype.getViewMatrix.

I think the headset projection matrices probably should not be touched, eg. used as is.

Also, I did not seem to get navigation with the Go controller to work. I think is supposed to do something. Maybe because the fish tank sphere is 80m in diameter.

It may be time to try live debugging on Android with a usb connected developer console ...

**Comment by Joo918** [comment]
I didn't push the Go controller navigation into GitHub yet because I couldn't resolve my unmatching displayId issue.
As for the viewpoint and viewarea, I've managed to get up to getting the previous active viewpoint (before VR started) position and setting the Viewarea position to that position, but I couldn't figure out  how to know when the VR button is clicked and user goes into VR. (Maybe an event triggered somewhere?)

**Comment by andreasplesch** [comment]
In the Viewarea code, the presence of this.vrFrameData is used to determine that VR is used to present.

https://github.com/x3dom/x3dom/blob/b90d5cfa2be1d67605cf4a36ad98e6a9cddfffb8/src/Runtime.js#L1329

shows how to check if in VR mode.

When the VR button div is clicked runtime.vrToggle is called which then checks.

Why not fork x3dom on github and share some experiments even if they do not work that well ?

**Comment by andreasplesch** [comment]
Sorry, hit the closing button by. accident.

**Comment by andreasplesch** [comment]
I made a webVR_viewpoint branch and added my idea above. It turned out that the gamepad (controller) navigation does an initial fit of the view which is responsible for spawning somewhere pretty far from he origin. I disabled that since viewpoints should really be responsible for setting initial views so the author has control. 
With these changes there is a more predictable view when changing into VR mode:

https://raw.githack.com/andreasplesch/x3dom/d43c165aa41140569244e6808732e4ce90f2fd85/test/functional/walkThrough.html

https://github.com/andreasplesch/x3dom/blob/webVR_viewpoint/src/Viewarea.js

Sofar only tested with the webvr emulator extension (still so useful). A side effect is that now mouse and keyboard also get applied which could be useful for cardboard without controllers.

I think the idea of the initial fit was to have a good view of a single model object by default in VR but it does not work well if the scene is large or designed for the avatar at the origin. I think a controller button should just reset the view to the active viewpoint without offsets, another then switch to the next viewpoint. Another good navigation pattern is to point at something with a laser beam and then move there by click.
Perhaps there could be new node or NavigationInfo field to map navigation actions to VR controller buttons so an author has some control.

**Comment by andreasplesch** [comment]
I got the remote Android debugging to work on the Galaxy S6 edge+.
There is an Uncaught TypeError after switching to VR:
```
VRControllerManager.js:179 Uncaught TypeError: Cannot read property 'axesScale' of undefined
    at x3dom.VRControllerManager.onUpdate (VRControllerManager.js:179)
    at x3dom.VRControllerManager.update (VRControllerManager.js:171)
    at x3dom.Viewarea.updateGamepads (Viewarea.js:180)
    at x3dom.X3DDocument.render (X3DDocument.js:372)
    at x3dom.X3DCanvas.tick (X3DCanvas.js:1217)
    at mainloop (X3DCanvas.js:1293)
```
![image](https://user-images.githubusercontent.com/6171115/54358670-4c5fc080-4637-11e9-86a6-e6a8393478bb.png)

It looks like it has to do with no controllers being available on cardboard. Perhaps this is related to the timeout issue.

**Comment by Joo918** [comment]
Oh! That makes a lot of sense!
I'll try to fix that and try it out!
Thank you!

**Comment by andreasplesch** [comment]
No, I fixed it on an Override but still get the timeout.

**Comment by andreasplesch** [comment]
Actually, this was just an issue with the debugger somehow. Checking for no controllers and therefore avoiding the Uncaught error, makes this scene now work for me on cardboard here:

https://raw.githack.com/andreasplesch/x3dom/e19d4b91ca425fd73d94c8292f39dcad4f655336/test/functional/walkThrough.html

**Comment by andreasplesch** [comment]
Let's see if I can try this on other devices as well.

**Comment by andreasplesch** [comment]
https://raw.githack.com/andreasplesch/x3dom/31a4c8290f3935bb25c8beb0943aeaf81c69bad5/test/functional/walkThrough.html

**Comment by Joo918** [comment]
Amazing! Thank you so much!
I've just tried your url on my Galaxy S8+ and it works.
I'll also try it out on Gear VR later tonight!

**Comment by andreasplesch** [comment]
For new Vive: https://raw.githack.com/andreasplesch/x3dom/28d4ff882b367086eb6319cafab94356da111a3f/test/functional/walkThrough.html

**Comment by andreasplesch** [comment]
On all the devices the background sphere is rendered inconsistently for left and right eye.

**Comment by Joo918** [comment]
+your patch works on Gear VR.

**Comment by andreasplesch** [comment]
https://raw.githack.com/andreasplesch/x3dom/461a6069ae0fc727fb0bd47d003ed3393767fb07/test/functional/inline-gltf.html

**Comment by Joo918** [comment]
Hello!
Thank you for these updates!
I cloned your webVR-viewarea branch and retried the fish-pool VR spheres in here:
http://metagrid2.sv.vt.edu/~joo918/fish_pool_VR/
(The texture of the sphere's broken, however... This have been an issue from a while ago)

I could get the VR mode to launch in cardboard.
I could also launch into VR mode from GearVR while the remote was not connected.
However, when I tried to go into VR after connecting the remote, it froze at the loading bar like before.

**Comment by andreasplesch** [comment]
These links were just a way to easily test on the Go.

What is the error in the console remote debugging the GearVR phone ?

Did you try the Go ?

**Comment by tphan25** [comment]
Hi, when I was testing with the Oculus Rift, I had the initial issue of how in VR mode, you would spawn outside the sphere. I initially fixed this by changing the fit function to set viewarea._movement to 0, which worked to place me at center of sphere; however, when I tried to do the same with your webVR-viewarea branch versio (since my version didn't work on mobile devices), it still does not allow me to spawn at the center of the sphere, but rather at a slight offset. I can't really pinpoint the exact cause of this but believe it has something to do with the changes in your viewarea.js changing the leftviewmatrix and rightviewmatrix calculations.

So basically, i want to be able to use a mobile device (and thus tried your fork with the webVR-viewarea) but also spawn at the center of a videosphere. i will try to look into changes too but if you could help that'd be greatly appreciated.

**Comment by andreasplesch** [comment]
Here is what currently happens in webVR-viewpoint branch when you enter VR by pressing the VR button:

- runtime.toggleVR is called
- runtime.enterVR is called
- VR display on HMD device is requested
- runtime.onEnterVR is called: this is a hook intended to be overwritten. By default it is doing nothing. In the gltf examples it is used to autofit an initial view (more or less).
- the actual, final viewmatrix for the shader uses the HMD provided left/rightviewmatrix but also multiplies in the viewmatrix of the current viewpoint of the scene (including its transforms if any).
- the viewpoint of the fish-pool has a position of 0 -1.6 0 whereas is the hemisphere seems to be centered at 0 0 0 . Could this account for the slight offset ? Try a viewpoint position of 0 0 0 ?

(hemiScreen_01.x3d probably should have a explicit texturecoordinate node, and seems to have a large radius if units are meters).

**Comment by Joo918** [comment]
I've just tested with the Go, and the texture's still broken, but going into VR works fine.
I can try to apply the patch from #938 and see if that fixes the spheres.
I'll also try to see what I get on the console on GearVR..

**Comment by andreasplesch** [comment]
The fish_pool videosphere is an IFS hemisphere, so the fixed Sphere node does not apply. Do you have a live example ? You could just use github.
Unfortunately, I do not have GearVR remote. I tried another bluetooh controller with the phone. It as detected as gamepad and I still could get into VR but the buttons had no effect.



I do not have a GearVR remote.
